compiler/rustc_codegen_ssa/src/back/metadata.rs

```diff
@@ -221,6 +221,9 @@ pub(crate) fn create_object_file(sess: &Session) -> Option<write::Object<'static
 
         file.set_macho_build_version(macho_object_build_version_for_target(sess))
     }
+    if binary_format == BinaryFormat::MachO {
+        file.set_subsections_via_symbols();
+    }
```

Comment:
It's not clear to me that this needs to be in `create_object_file`? Maybe it would be better to only use it in `add_linked_symbol_object`?

Also, `MH_SUBSECTIONS_VIA_SYMBOLS` is vastly under-documented, I'd really like to see a comment here explaining why it's safe for us to use.

Reply:
[Object::set_subsections_via_symbols](https://docs.rs/object/latest/object/write/struct.Object.html#method.set_subsections_via_symbols) says it should be called before `add_section` or `add_subsection`, so I feel it better to put it here (not error-prone).

I don't know whether it's safe or not, either. I searched on the Internet and found that this is the only way for Mach-O to implement GC.

Reply:
My problem with having it in `create_object_file` is that it may negatively affect the other object files we create (which again is hard to tell for sure, since the docs around it are so limited).

Reply:
Ah! I didn't notice that this function has multiple callers.

I will add a parameter to control this behavior.

Reply:
Eh, I'd still have preferred to just change `add_linked_symbol_object`, it'd be unexpected if `create_object_file` added any sections or subsections, it's literally prefixed "create" as in "only create, do not fill".